ledger/src/shred/payload.rs

```diff
@@ -1,62 +1,61 @@
-use std::{
-    ops::{Deref, DerefMut},
-    sync::Arc,
+use {
+    bytes::{Bytes, BytesMut},
+    std::{
+        mem,
+        ops::{Bound, Deref, DerefMut, RangeBounds, RangeFull},
+        slice::SliceIndex,
+    },
 };
 
 #[derive(Clone, Debug, Eq)]
-pub enum Payload {
-    Shared(Arc<Vec<u8>>),
-    Unique(Vec<u8>),
+pub struct Payload {
+    pub bytes: Bytes,
 }
 
-macro_rules! make_mut {
-    ($self:ident) => {
-        match $self {
-            Self::Shared(bytes) => Arc::make_mut(bytes),
-            Self::Unique(bytes) => bytes,
-        }
-    };
-}
-
-macro_rules! dispatch {
-    ($vis:vis fn $name:ident(&self $(, $arg:ident : $ty:ty)?) $(-> $out:ty)?) => {
-        #[inline]
-        $vis fn $name(&self $(, $arg:$ty)?) $(-> $out)? {
-            match self {
-                Self::Shared(bytes) => bytes.$name($($arg, )?),
-                Self::Unique(bytes) => bytes.$name($($arg, )?),
-            }
-        }
-    };
-    ($vis:vis fn $name:ident(&mut self $(, $arg:ident : $ty:ty)*) $(-> $out:ty)?) => {
-        #[inline]
-        $vis fn $name(&mut self $(, $arg:$ty)*) $(-> $out)? {
-            make_mut!(self).$name($($arg, )*)
-        }
+impl Payload {
+    /// Convert the payload's inner [`Bytes`] into a [`BytesMut`], consuming the [`Payload`].
+    ///
+    /// If the payload is unique (single reference), this will return a [`BytesMut`] with the
+    /// contents of the payload without copying. If the payload is not unique, this will make a copy
+    /// of the payload in a new [`BytesMut`]. As such, take care to avoid performing this conversion
+    /// if the payload is not unique.
+    #[inline]
+    pub fn into_bytes_mut(self) -> BytesMut {
+        self.bytes.into()
     }
-}
 
-impl Payload {
+    /// Get a mutable reference via [`PayloadMutGuard`] to the payload's _full_ inner bytes.
+    /// See [`Payload::get_mut`] for selecting a subset of the payload's inner bytes.
+    ///
+    /// If the payload is unique (single reference), this will not perform any copying. Otherwise it
+    /// will. As such, take care to avoid performing this conversion if the payload is not unique.
     #[inline]
-    pub(crate) fn resize(&mut self, size: usize, byte: u8) {
-        if self.len() != size {
-            make_mut!(self).resize(size, byte);
-        }
+    pub fn as_mut(&mut self) -> PayloadMutGuard<'_, RangeFull> {
+        PayloadMutGuard::new(self, ..)
     }
 
     #[inline]
-    pub(crate) fn truncate(&mut self, size: usize) {
-        if self.len() > size {
-            make_mut!(self).truncate(size);
+    /// Get a mutable reference via [`PayloadMutGuard`] to a subset of the payload's inner bytes.
+    ///
+    /// If the payload is unique (single reference), this will not perform any copying. Otherwise it
+    /// will. As such, take care to avoid performing this conversion if the payload is not unique.
+    pub fn get_mut<I>(&mut self, index: I) -> Option<PayloadMutGuard<'_, I>>
```

Comment:
Is it wise to have `as_mut` and `get_mut` methods that might secretly Clone under the hood? `get_mut` usually implies a fast-ish getter method rather than something that might clone under the hood. I'd rather work with `try_get_mut` that would error when another reference exists, preventing accidental creation of copies. And if someone is really happy with a copy made, they can explicitly call `.clone()` followed up by `try_get_mut().unwrap()`.  What do you think?

Reply:
Yeah I see your point. Either way making `as_mut` do allocations sounds like a massive antipattern. Can we name it otherwise so it is explicitly separated from https://doc.rust-lang.org/std/convert/trait.AsMut.html ? Like, `get_mut_or_clone` or whatever to imply that this can get expensive?

Reply:
> And the practical reality is that payloads should be unique while we're
> building them

Can we assert then? I agree that Result is likely overkill, but doing implicit
copies accidentally would also be bad.

It's frustrating because the code that builds shreds is so involved, it should
be refactored to just build into BytesMut. That doesn't seem like a change for
this PR tho, nor necessarily something that we should be doing in the perf team. @alexpyattaev this sounds more like something you should have in the net
team backlog.

Reply:
This is fine, let us take the small win for now. If https://github.com/anza-xyz/agave/pull/7413 lands before 3.0 we will be able to make any changes in shred module without care for API stability.